### PR TITLE
Add timeout parameter.

### PR DESCRIPTION
**Pull request by tillahoffmann**
#### Submission Checklist

- [x] Run unit tests (on macOS; will have to wait for Windows and Linux in CI)
- [x] Declare copyright holder and open-source license: see below

#### Summary

This PR adds a `timeout` parameter to `optimize`, `sample`, `generate_quantities`, and `variational` including unit tests. I was running hyperparameter sweeps for a model and some configurations took much longer than others. Setting a timeout means that "bad" configurations can be aborted if they take too long.

I used a `threading.Timer` to terminate the cmdstan subprocess. Checking the elapsed time in `_run_cmdstan` didn't work because the `readline` call blocks until the next line of `stdout` becomes available (which can take a while if the model is sampling inefficiently).

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

Harvard University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

File: test/test_variational.py [hunk 1]
```diff
@@ -291,6 +291,14 @@ def test_attrs(self):
         with self.assertRaisesRegex(AttributeError, 'Unknown variable name:'):
             dummy = fit.c
 
+    def test_timeout(self):
+        stan = os.path.join(DATAFILES_PATH, 'timeout.stan')
+        timeout_model = CmdStanModel(stan_file=stan)
+        self.assertRaisesRegex(
```
Comment (WardBrian) [inline test/test_variational.py:297]: Would you mind reorganizing these tests so they use the `with self.assertRaises...` style?

File: cmdstanpy/model.py [hunk 2]
```diff
@@ -1579,6 +1596,12 @@ def _run_cmdstan(
                 env=os.environ,
                 universal_newlines=True,
             )
+            if timeout:
+                timer = threading.Timer(timeout, proc.terminate)
```
Comment (WardBrian) [inline cmdstanpy/model.py:1600]: Should we later be calling `timer.cancel()` if the task finishes before this timeout hits?

**Review by WardBrian — CHANGES_REQUESTED** [review]
This is a nice idea, thanks for looking into it. 

Do you have the `black` formatter installed? Some of the changes look like they need to be formatted

File: cmdstanpy/model.py (hunk 2 above)
Comment (tillahoffmann) [reply cmdstanpy/model.py:1600]: Ah neat, didn't know about that function.

**Comment by tillahoffmann** [comment]
I just raised `RuntimeError` for the time being. Thoughts on raising `TimeoutError` instead so it's easier to catch the exception?

**Comment by WardBrian** [comment]
That would certainly be clearer, but I assume it has some difficulties with our `check_returncode`/`get_err_msg` functionality?

**Comment by tillahoffmann** [comment]
I'm currently manually patching the return code to 60 as `os.strerror(60) == 'Operation timed out'` if the timer terminates the subprocess. So we could raise a `TimeoutError` based on the return code?

**Comment by WardBrian** [comment]
Sounds good to me

**Comment by WardBrian** [comment]
I suppose the other thing you could do besides patch the return code is have the Timer function both terminate the proc _and_ indicate it did this by setting an attribute in runset. I'm mostly indifferent, especially if we think that `-15` as a return code would never show up normally

**Comment by tillahoffmann** [comment]
I've tried it with storing state in `runset` which might also fix the failing Windows tests.

**Comment by tillahoffmann** [comment]
Alright, these seem to be passing now.

As a timing-related thing, what are your thoughts on adding a `time.sleep(1e-3)` in the `while proc.poll() is None` loop [here](https://github.com/stan-dev/cmdstanpy/pull/621/files#diff-af39d3679809ef0504a5e945eb5588b237ae0c42c92631c0c2f626b14edb414dL1582-L1591)? That would reduce the CPU overhead from the busy waiting a bit. The `subprocess.Popen.wait` method [uses a sleep of 1 ms](https://github.com/python/cpython/blob/8a221a853787c18d5acaf46f5c449d28339cde21/Lib/subprocess.py#L1961), for example.

**Comment by WardBrian** [comment]
Seems reasonable, but looking at that code it seems like it will do the wrong thing if more than one line of output is buffered at a time (which is bad)? So we would also definitely want to do `readlines` instead of `readline`, and process any potential output in a loop, I think.

**Review by WardBrian — APPROVED** [review]
This looks good to me. Do you want to look at the proc.poll issue in a different PR? If so I will merge this

**Comment by tillahoffmann** [comment]
Makes sense; maybe I'll have a look at that in a separate PR.

**Comment by tillahoffmann** [comment]
As a follow up, the issue of busy-waiting is not an issue: `readline` blocks when there's no content so the loop is only busy when there's actually stuff to do.

**Pull request merged**